Add a coder to reread syntaxtrees

sebasguts opened the pull request:
See [here](https://github.com/fingolfin/gap/pull/114) for the initial discussion and history

fingolfin commented:
The first commit "WIP: add SyntaxTreeDefaultCoder" should be removed.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -236,15 +333,66 @@ static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
+{
+    Obj  value = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 0));
+    Obj  string = GET_VALUE_FROM_CURRENT_BODY(READ_EXPR(expr, 1));
+    UInt char_val = (UInt)READ_EXPR(expr, 2);
```
```suggestion
    UChar chr = (UChar)READ_EXPR(expr, 2);
```
Otherwise an out of bounds access could happen a few lines later on.

And `char_val` violates our naming conventions.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -236,15 +333,66 @@ static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
+    UInt char_val = (UInt)READ_EXPR(expr, 2);
+    AssPRec(result, RNamName("value"), value);
+    AssPRec(result, RNamName("string"), string);
+    AssPRec(result, RNamName("mark"), ObjsChar[char_val]);
```
```suggestion
    AssPRec(result, RNamName("mark"), ObjsChar[chr]);
```

sebasguts commented on src/code.c:
```diff
@@ -1909,7 +1893,10 @@ static void CodeLazyFloatExpr(Obj str)
+    if(pushExpr) {
+        PushExpr(fl);
+    }
+    return fl;
```
I really do not understand this comment. All modifications in `CodeLazyFloatExpr` are made in `0a00459bd2a3eb87df8d0ef22561c1d7a2d770e5`, and then the next commit uses it.

sebasguts replied:
I now understand this comment. Apparently, Github displays commits by date, not in the order they appear on the branch . . . stupid rebasing. Should I make a rebase with new dates?

ChrisJefferson replied:
adding `--ignore-date` to rebasing will make it set all the dates to the current date, which gets all your commits in the "right order" again.

sebasguts replied:
Yeah, but I tent to not do that, for reasons that are probably not really logical. Will rebase this one with new dates.

sebasguts replied:
Did it, now they all have the same date and still appear in the wrong order . . .

sebasguts commented:
@fingolfin I think I adressed all your comments. I also added sensible error messages for checking if a needed entry in a syntax tree node record is bound.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,92 @@
+    if (IsbPRec(typeRec, rnam)) {
+        return (UInt1)UInt_ObjInt(ElmPRec(typeRec, rnam));
+    }
+    ErrorQuit("Unrecognized syntax tree node type %s",
```
```suggestion
    ErrorQuit("Unrecognized syntax tree node type %g",
```

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,92 @@
+        return (UInt1)UInt_ObjInt(ElmPRec(typeRec, rnam));
+    }
+    ErrorQuit("Unrecognized syntax tree node type %s",
+              (Int)CSTR_STRING(typestr), 0);
```
```suggestion
              (Int)typestr, 0);
```

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,92 @@
+    if (!IsbPRec(node, type)) {
+        ErrorQuit("while coding: <type> field is not present", 0, 0);
+    }
+    return FIND_TNUM_FROM_STRING(ELM_REC(node, type));
```
```suggestion
    return FIND_TNUM_FROM_STRING(ElmPRec(node, type));
```
You just called `IsbPRec` a few lines before, and use `ElmPRec`, so might as well use that here. Or, if we truly want to support general records (in case somebody ever adds some), we should be using `ISB_REC`, `UNB_REC`, `ELM_REC` everywhere. But I think it's OK to restrict to precs

sebasguts commented:
This PR is now based on #3370

fingolfin commented on src/code.h:
```diff
@@ -109,6 +109,25 @@ EXPORT_INLINE Obj VALUES_BODY(Obj body)
+**  and file when NewStat was called is used. line=0, file=0 is used
+**  to denote a statement which should not be tracked.
+*/
+Stat NewStatWithProf(UInt type, UInt size, UInt line);
```
Better comment:
```
/****************************************************************************
**
*F  NewStatOrExpr(<type>,<size>,<line>,<file>) . . . allocate a new statement
**
**  'NewStatOrExpr' allocates a new statement or expressions memory block of
**  type <type> and with <size> bytes. It also records the line number <line>
**  and file id <file> as source of the statement for profiling. It returns
**  the offset of the new statement.
**
**  Callers may pass zero for both <line> and <file> to denote a statement
**  which should not be tracked by the profiling code.
*/
```
and then rename `NewStatWithProf` to `NewStatOrExpr` globally (could be in a commit of its own, don't care).

fingolfin commented on src/code.h:
```diff
@@ -661,12 +684,25 @@ void CodeFuncCallEnd(UInt funccall, UInt options, UInt nr);
+*F  AddValueToBody( <val> ) . . . . . . . . . . .  store value in values list
+**
+**  'AddValueToBody' adds a value into the value list of the body of the
+**  function currently being coded, and returns the index  at which the
```
```suggestion
**  function currently being coded, and returns the index at which the value
```

fingolfin commented on src/code.h:
```diff
@@ -661,12 +684,25 @@ void CodeFuncCallEnd(UInt funccall, UInt options, UInt nr);
+**
+**  'AddValueToBody' adds a value into the value list of the body of the
+**  function currently being coded, and returns the index  at which the
+**  value was inserted. This function must only be called while coding a
```
```suggestion
** was inserted. This function must only be called while coding a function.
```

fingolfin commented on src/code.h:
```diff
@@ -661,12 +684,25 @@ void CodeFuncCallEnd(UInt funccall, UInt options, UInt nr);
+**  'AddValueToBody' adds a value into the value list of the body of the
+**  function currently being coded, and returns the index  at which the
+**  value was inserted. This function must only be called while coding a
+**  function.
```
```suggestion
```

fingolfin commented on src/code.h:
```diff
@@ -661,12 +684,25 @@ void CodeFuncCallEnd(UInt funccall, UInt options, UInt nr);
+**  function currently being coded, and returns the index  at which the
+**  value was inserted. This function must only be called while coding a
+**  function.
+**
```
```suggestion
```

fingolfin commented on src/code.h:
```diff
@@ -1069,6 +1105,16 @@ void CodeListExprEnd(UInt nr, UInt range, UInt top, UInt tilde);
+
+/****************************************************************************
+**
+*F  CodeLazyFloatExpr(<str>,<pushExpr>) . . .  code literal string expression
```
```suggestion
*F  CodeLazyFloatExpr(<str>,<pushExpr>) . . . . .  code lazy float expression
```

fingolfin commented on src/scanner.c:
```diff
@@ -746,6 +746,30 @@ static void GetStr(void)
+    UInt i = 0;
+    Char c = PEEK_CURR_CHAR();
+
+    while ( c != '\n' && c != '\r' && c != '\f' && c != '\t' && c != '\377') {
```
```suggestion
    while (c != '\n' && c != '\r' && c != '\f' && c != '\t' && c != '\377') {
```

fingolfin commented on src/scanner.h:
```diff
@@ -63,6 +63,8 @@ enum SCANNER_SYMBOLS {
     S_TILDE             = (1UL<<11)+4,
     S_HELP              = (1UL<<11)+5,
+    S_PRAGMA            = (1UL<<11)+6,
+
```
```suggestion
```

fingolfin commented on src/syntaxtree.c:
```diff
@@ -272,6 +453,57 @@ static Obj SyntaxTreeIf(Obj result, Stat stat)
+        Obj  conditionnode = ElmRecST(tnum, condbodypair, "condition");
+        Obj  bodynode = ElmRecST(tnum, condbodypair, "body");
+        Expr condition = SyntaxTreeDefaultExprCoder(conditionnode);
+        Expr body = SyntaxTreeDefaultStatCoder(bodynode);
```
```suggestion
        Stat body = SyntaxTreeDefaultStatCoder(bodynode);
```

fingolfin commented on src/syntaxtree.c:
```diff
@@ -236,15 +366,66 @@ static Obj SyntaxTreeRecExpr(Obj result, Expr expr)
-    Obj string;
+    UInt1 tnum = GetTypeTNum(node);
+    Obj   keyvalue = ElmRecST(tnum, node, "keyvalue");
+    UInt  len = LEN_LIST(keyvalue);
```
```suggestion
    UInt  len = LEN_PLIST(keyvalue);
```

ChrisJefferson replied:
I might be going the other way to @fingolfin, but I was a bit worried at the number of uses of the PLIST functions in this, which don't check they are PLISTs. Using the LIST functions instead, while a little slower, would at least make sure GAP didn't crash/corrupt.

sebasguts replied:
I wouldn't mind using either the one or the other.

sebasguts commented:
I think I addressed all your comments, and also added the checks we discussed today.

Should we try to add the context object, too, or would be a different PR a better place.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -279,6 +482,49 @@ static Obj SyntaxTreeCompilePragma(Obj result, Stat stat)
+{
+    UInt1 tnum = GetTypeTNum(node);
+    Obj   branches = ElmRecST(tnum, node, "branches");
+    UInt  len = LEN_PLIST(branches);
```
So @ChrisJefferson pointed out in another comment (now on "outdated" code, so not immediately visible), where I suggested to change a `LEN_LIST` to `LEN_PLIST`, that we could go the other way, and change all PLIST functions to generic ones, for safety. I am also fine with this -- the alternative being to add lost of "is this really a pist" checks everywhere. What I do not like is code which calls `LEN_LIST` and then immediately `ELM_PLIST`, as this code used to do, and as unfortunately still quite some other kernel code does. It combines the worst of both worlds: inferior performance plus a risk of crashes...

Anyway, performance shouldn't matter much here, so I'd be fine if we replaced `LEN_PLIST` with `LEN_LIST`, `ELM_PLIST` with `ELM_LIST` (resp. `ELM0_LIST` in the few cases where it is needed), `SET_ELM_PLIST` with `ASS_LIST`, etc.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,101 @@
+static Obj ElmRecST(UInt1 tnum, Obj node, const char * name)
+{
+    UInt rnam = RNamName(name);
+    if (!IsbPRec(node, rnam)) {
```
We never verify that the object we get is a precord, do we? So this would crash if I created malicious input (respectively: if I try to create a valid syntax tree object but screwed it up).

As with lists, we can either insert `RequirePRecord` tests everywhere (after creating such a function). Or we just use `ISB_REC`, `ELM_REC`, etc.

It might also be instructive to write malicious input just to see how it crashes and burns w/o these error checks. :-).

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,101 @@
 
-#define ARGS(name) ARG(name, 0)
+#define ARG_EXPR(name, func, coder) ARG(name, func, coder, 0)
+// The corresponding ARG_STAT is not used
```
Huh? Perhaps you meant this?
> There is no ARG_STAT, because it is never used

That surprise me, though: T_ATOMIC contains a single body statement, plus a variable list of expressions for locks... Hmm, looking at your code for T_ATOMIC, it seems to be incorrect, which would explain why there is no ARG_STAT here.

sebasguts replied:
The single body statement is handles by `ARG_STAT_` since it is just a `T_SEQ...` I think. So we do not need `ARG_STAT` which allows for a custom coder/compiler.

sebasguts replied:
Anyway, the code was indeed wrong. Fixed that.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,101 @@
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
 
-#define ARGS(name) ARG(name, 0)
+#define ARG_EXPR(name, func, coder) ARG(name, func, coder, 0)
```
Now would be a good time to also add some comments that explain what these macros do... What is `ARG_EXPR` about? Perhaps like this:
> // ARG_EXPR indicates an "expression argument", i.e., that the corresponding slot of the
> // statement or expression type it belongs to is meant to reference an expression.
> // This information is used for validation when coding an AST back into a T_BODY object.
> // The expression is stored in resp. read from the record entry <name>.

And then later:
> // ARG_STAT is like ARG_EXPR, except that it indicates that the corresponding slot should contain
> // a statement instead of an expression.

It's not great, but perhaps better than nothing. And it could be adapted to A

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,101 @@
+    ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultExprCoder, 0)
 
+#define ARGS_STAT(name) ARG(name, 0, 0, 1)
+#define ARGS_EXPR(name) ARG(name, 0, 0, 0)
```
> // ARGS_EXPR can be used instead of ARG_EXPR as the last argument to the COMPILER_
> // macro, and indicates that a variable number of arguments follow; all are supposed to be
> // expressions; and all of them are to be stored into a plist which is stored in resp. read
> // from the record entry <name>.

fingolfin commented on src/syntaxtree.c:
```diff
@@ -31,50 +31,101 @@
 
+static Obj typeStrings;    // plain list mapping statement tnums to strings
+                           // like "T_INTEXPR"
+static Obj typeRec;    // plain record mapping strings like "T_INTEXPR" to the
```
We usually put such comments before the variable, not after it, as you do here.

sebasguts commented:
@fingolfin I think I adressed all of your comments

sebasguts commented:
@ChrisJefferson @fingolfin ping?

ChrisJefferson commented on src/code.h:
```diff
@@ -109,6 +109,25 @@ EXPORT_INLINE Obj VALUES_BODY(Obj body)
+**  and file id <file> as source of the statement for profiling. It returns
+**  the offset of the new statement.
+**
+**  Callers may pass zero for both <line> and <file> to denote a statement
```
I agree. This is an old bug, at some point the file argument disappeared and the comment wasn't fixed, but it would be reasonable to fix it now.

sebasguts commented on src/syntaxtree.c:
```diff
@@ -31,56 +31,127 @@
+
+
+// ARG_EXPR_ can be used instead of ARG_EXPR if the default compiler
+// and coder should be used to compile the corresponding expression or code
```
I am fine with using any of those terms. `Code` is currently used in the interface, `Compile` is not. So changing compile to `decode` is the easiest. But we can safely assume that there is nobody using it right now anyway, so we can just change it to whatever we want.

sebasguts commented:
I think I adressed all the comments now :)

ChrisJefferson commented:
So, possibly stupid question, is this supposed to be complete, or progress towards a rereader?

I tried extending the test which tests SYNTAX_TREE on every function, by the following horrible patch:

```
@@ -25,11 +25,17 @@ gap> for n in NamesGVars() do
 >            v := ValueGlobal(n);
 >            if IsFunction(v) and not IsKernelFunction(v) then
 >                SYNTAX_TREE(v);
+>                        if not test_tree(v) then
+>                           Print("failed round trip: ",v,"\n");
+>                        fi;
 >            elif IsOperation(v) then
 >                for i in [1..6] do
 >                    for x in METHODS_OPERATION(v, i) do
 >                        if IsFunction(x) and not IsKernelFunction(v) then
 >                        SYNTAX_TREE(x);
+>                        if not test_tree(x) then
+>                           Print("failed round trip: ",v,"\n");
+>                        fi;
 >                        fi;
 >                    od; 
 >                od;
```

And GAP segfaults.

sebasguts commented:
@ChrisJefferson It does not work on operations, no. But I should really check that when calling `SyntaxTree` The segfault is really bad . . .

sebasguts commented:
Wait, this is a different issue. Sorry.

ChrisJefferson commented:
This is enough to upset it:

```
gap> f := function() return [,[]]; end;
function(  ) ... end
gap> SYNTAX_TREE_CODE(SYNTAX_TREE(f));
```

sebasguts commented:
I will fix that. I am not sure how that slipped. Thank you.

Anyway, there is a thing that I am aware of that does not work yet, namely `HVAR` references that do not point to the "read" function context anymore, but outside.

I plan to overcome this by adding a `T_REF_VALUE` (and `T_ASS_VALUE`) to GAP, so we can store those references not as `HVAR`s anymore, but as values directly. However, this of course may change the overall semantics of a function . . .

ChrisJefferson commented:
(I don't know if this is reasonable), if in cases where SYNTAX_TREE_CODE failed it just returned 'fail', then we could obviously test for when it works, or returns fail, and then still loop over all the functions? (I don't know if that's reasonable given the way it works internally)

sebasguts commented:
@ChrisJefferson Currently, they throw an error, which I would guess is more reasonable. But I could do a `no_error` version for testing, if you would prefer?

The problem rereading those `HVAR` functions is that rereading them works, but their result is arbitrary, see here (which is worse than a segfault)):
```
gap> xx := (i -> ( j-> i + j ) )(3);
function( j ) ... end
gap> SYNTAX_TREE(SYNTAX_TREE_CODE(SYNTAX_TREE(xx)))=SYNTAX_TREE(xx);
true
gap> SYNTAX_TREE_CODE(SYNTAX_TREE(xx))(3);
Z(3)
```

ChrisJefferson commented:
I was just thinking about if it would be reasonable to merge something like the diff I made, to test a large set of functions. If we are still too far away from that, we can also merge the current partial code, and work on improving it later.

sebasguts commented:
No, absolutely reasonable. I will include it.

sebasguts commented:
@ChrisJefferson I added the full cycle test, and they pass
@fingolfin I made additional changes to `CodeFuncExprEnd`, please have a new look at the changes there.

fingolfin commented on tst/testinstall/syntaxtree.tst:
```diff
@@ -10,19 +10,30 @@ gap> SyntaxTree(x -> x);
 >                        if IsFunction(x) and not IsKernelFunction(v) then
->                        SYNTAX_TREE(x);
+>                            if not test_tree(v) then
+>                                Print("failed round trip: ",n,"\n");
```
Would be more useful like this:
```suggestion
>                                Print("METHODS_OPERATION(", n, ",", i, ") failed round trip\n");
```

fingolfin commented on src/code.c:
```diff
@@ -874,11 +854,21 @@ void CodeFuncExprEnd(UInt nr)
-        {
-            CodeReturnVoidWhichIsNotProfiled();
-            nr++;
+        while( 1 ){
```
Code formatting!
```suggestion
        while (1) {
```

fingolfin replied:
Also, this code *really* needs to be commented. Both what it is doing (see my suggestion below), but also more importantly, *why* it is doing it. Which I don't know right now...

fingolfin replied:
Why doesn't it suffice to insert the following into the existing code, to achieve the same effect?
```
  // find the last statement
  while (T_SEQ_STAT <= TNUM_STAT(stat1) && TNUM_STAT(stat1) <= T_SEQ_STAT7) {
    TNUM_STAT(stat1) <= T_SEQ_STAT7) {
    UInt size = SIZE_STAT(stat1) / sizeof(Stat);
    stat1 = READ_STAT(stat1, size - 1);
  }
```

sebasguts replied:
Well, logically, this does exactly the same, except it is not intertwined with the addition of the `return void` to the statement.

sebasguts replied:
About the why: The thing is that if we reread a syntax tree, the statements are already nested, i.e., if there are more than 7 statements in the body, the last one is already a new `SEQ_STAT*`. This ensures that the `return;` is only added if there is really no `return` in the last statement.

When a function is coded, the statements are packed into nested statements later.

I discovered this problem when I realized that "long" functions get an additional `return;`
